### PR TITLE
fix(flagship fighters): Proper carry and handling for ex-flagship fighters

### PR DESCRIPTION
**Pull request by petervdmeer**
**Bugfix:** This PR addresses issue #6519

## Fix Details
Fighters that were unloaded will now properly show up on launch again because UnMarkForRemoval() is added to UnloadBays(). Thanks to the [analysis by vitalchip](https://github.com/endless-sky/endless-sky/issues/6519#issuecomment-1025215292).

If there are zero bays available for some carried ship category, then the on-launch-warning for such carried ships not being able to jump will still show up. (Was reported as part of #6519, but was actually a separate issue.)

The Ship::AllowCarried() function got replaced by checks if the relevant ship is a flagship. (Since being a flagship is not a property of a ship, but a property of the players fleet.) This fix uses quite a bit of code originally created for #6323 (thanks to Zitchas for noticing the commonality).

Fixes: https://github.com/endless-sky/endless-sky/issues/6519

## Testing Done
 Reviewed all locations (AI, Engine, Fleet, HailPanel, MapPanel, MissionAction, Mission, NPC, PlayerInfo, Ship) where CanBeCarried() was used to check if any of those need an exception for flagships.
 Performed the test-steps as described in #6519 (not with the savegame provided in #6519, but with a Bactrian carrier and a fast Finch).

## Save File
See #6519 for a savegame to test for the bug (and to test this fix).

### CODE REVIEW

File: source/Ship.cpp [hunk 1]
```diff
@@ -3536,6 +3536,7 @@ void Ship::UnloadBays()
 			carriedMass -= bay.ship->Mass();
 			bay.ship->SetSystem(currentSystem);
 			bay.ship->SetPlanet(landingPlanet);
+			bay.ship->UnmarkForRemoval();
```
Comment (Hurleveur) [inline source/Ship.cpp:3539]: Not sure why this is required?

**Review by Hurleveur — CHANGES_REQUESTED** [review]
Looks good overall, and like the possibility to have a fighter flagship wasnt refined enough for it to function.
Only are a few quick fixes but I dont see anything wrong with the logic (I cant playtest it currently though)

File: source/Ship.cpp (hunk 1 above)
Comment (petervdmeer) [reply source/Ship.cpp:3539]: There is a nice [analysis by Vitalchip](https://github.com/endless-sky/endless-sky/issues/6519#issuecomment-1025215292) in the reporting issue on why this is required.

Comment (Hurleveur) [reply source/Ship.cpp:3539]: > Addressing point 3 - Deploying and recalling the Finch sets "shouldRemove" but this isn't unset when unloading bays, only when launching.

Seems to me lile we should instead unset it when unloading bays but I dont know this code in details so take my advice with a grain of salt.

Comment (petervdmeer) [reply source/Ship.cpp:3539]: Well, this is in the `Ship::UnloadBays()` function, so it looks like we are now actually doing exactly what you suggest.

We might still be able to optimize if we take the common code from Launch() and from UnloadBays() and place such common code on 1 single location. But let's keep that for another PR.

File: source/Engine.cpp [hunk 2]
```diff
@@ -1632,7 +1632,9 @@ void Engine::MoveShip(const shared_ptr<Ship> &ship)
 
 	// Boarding:
 	bool autoPlunder = !ship->IsYours();
-	shared_ptr<Ship> victim = ship->Board(autoPlunder);
+	shared_ptr<Ship> victim;
+	if(ship.get() != flagship || !ship->CanBeCarried())
+		victim = ship->Board(autoPlunder);
 	if(victim)
 		eventQueue.emplace_back(ship, victim,
 			ship->GetGovernment()->IsEnemy(victim->GetGovernment()) ?
```
[the remark this answers is not in the document]
Comment (petervdmeer) [reply source/Engine.cpp:1640]: True, we can keep the victim very local. I'll update the code for this.

File: source/PlayerInfo.cpp [hunk 3]
```diff
@@ -49,6 +49,26 @@ PARTICULAR PURPOSE.  See the GNU General Public License for more details.
 
 using namespace std;
 
+namespace {
+	// Move the flagship to the start of your list of ships. It does not make
+	// sense that the flagship would change if you are reunited with a different
+	// ship that was higher up the list. This function should not be used while
+	// other functions have direct references to the shared_pointer elements in
+	// this list.
```
[the remark this answers is not in the document]
Comment (petervdmeer) [reply source/PlayerInfo.cpp]: Yes, I added a newline between the different subjects.

File: source/PlayerInfo.cpp [hunk 4]
```diff
@@ -49,6 +49,26 @@ PARTICULAR PURPOSE.  See the GNU General Public License for more details.
...
 
+namespace {
+	// Move the flagship to the start of your list of ships. It does not make sense
+	// that the flagship would change if you are reunited with a different ship that
+	// was higher up the list.
+	// This function should not be used while other functions have direct references
+	// to the shared_pointer elements in this list.
+	void MoveFlagshipBegin(vector<shared_ptr<Ship>> &ships, const shared_ptr<Ship> &flagship)
+	{
+		if(flagship)
```
Comment (quyykk) [inline source/PlayerInfo.cpp:60]: Use early return instead of a big (ok it's not that big but still).

File: source/PlayerInfo.cpp [hunk 5]
```diff
@@ -795,6 +815,45 @@ const shared_ptr<Ship> &PlayerInfo::FlagshipPtr()
 
 
 
+// Set the flagship (on departure or during flight).
+void PlayerInfo::SetFlagship(Ship &other)
+{
+	// Remove active data in the old flagship.
+	if(flagship && flagship.get() != &other)
```
Comment (quyykk) [inline source/PlayerInfo.cpp:822]: Can't we return early if the flagship is already `other`? Or guard the call to MoveFlagshipBegin and SetFlagship.

File: source/PlayerInfo.cpp [hunk 6]
```diff
@@ -49,6 +49,26 @@ PARTICULAR PURPOSE.  See the GNU General Public License for more details.
...
+	// to the shared_pointer elements in this list.
+	void MoveFlagshipBegin(vector<shared_ptr<Ship>> &ships, const shared_ptr<Ship> &flagship)
+	{
+		if(flagship)
+		{
+			auto it = find(ships.begin(), ships.end(), flagship);
+			if(it != ships.begin() && it != ships.end())
+			{
+				ships.erase(it);
+				ships.insert(ships.begin(), flagship);
```
Comment (quyykk) [inline source/PlayerInfo.cpp:66]: This is a bit inefficient since it involves moving every element in the vector. You can use [`std::move_backward`](https://en.cppreference.com/w/cpp/algorithm/move_backward) to only move every ship (saving + setting the first element to the saved flagship after the move).

File: source/Ship.cpp [hunk 7]
```diff
@@ -2969,6 +2961,22 @@ void Ship::WasCaptured(const shared_ptr<Ship> &capturer)
 
 
 
+// Clear all orders and targets this ship has (after capture or transfer of control).
+void Ship::ClearTargetsAndOrders()
+{
+	commands.Clear();
```
Comment (quyykk) [inline source/Ship.cpp:2967]: Clear `firingCommands` as well, this was an oversight on my party. oops

File: source/PlayerInfo.cpp (hunk 5 above)
Comment (petervdmeer) [reply source/PlayerInfo.cpp:822]: No, it is possible that some on-planet actions already resulted in `flagship.get() == &other` being true, while some of the other actions were done yet. (The on-planet flagship handling only looks at positions in the ships list.)

Also, the player could have unparked some ships just before launch, those ships also need to be updated on line 831.

**Review by Hurleveur — DISMISSED** [review]

File: source/PlayerInfo.cpp [hunk 8]
```diff
@@ -49,6 +49,26 @@ PARTICULAR PURPOSE.  See the GNU General Public License for more details.
...
+	// to the shared_pointer elements in this list.
+	void MoveFlagshipBegin(vector<shared_ptr<Ship>> &ships, const shared_ptr<Ship> &flagship)
+	{
+		if(!flagship)
+			return;
+
+		auto it = find(ships.begin(), ships.end(), flagship);
+		if(it != ships.begin() && it != ships.end())
+		{
+			move_backward(ships.begin(), prev(it), it);
```
Comment (quyykk) [inline source/PlayerInfo.cpp:66]:
```suggestion
			move_backward(ships.begin(), it, next(it));
```

The range doesn't include the last element. As such the end is `it`. And the third argument is the new end, i.e. one past the actual end so we need to increment it by one.

Comment (petervdmeer) [reply source/PlayerInfo.cpp]: Sorry, I should have read the documentation better.

File: source/Engine.cpp [hunk 9]
```diff
@@ -1632,11 +1632,14 @@ void Engine::MoveShip(const shared_ptr<Ship> &ship)
 
 	// Boarding:
 	bool autoPlunder = !ship->IsYours();
-	shared_ptr<Ship> victim = ship->Board(autoPlunder);
-	if(victim)
-		eventQueue.emplace_back(ship, victim,
-			ship->GetGovernment()->IsEnemy(victim->GetGovernment()) ?
-				ShipEvent::BOARD : ShipEvent::ASSIST);
+	if(ship.get() != flagship || !ship->CanBeCarried())
```
Comment (quyykk) [inline source/Engine.cpp:1635]: This makes it so that if you are in a fighter you can't board anymore. Why is this change necessary?

Comment (petervdmeer) [reply source/Engine.cpp:1635]: The BOARD command for fighters is also used for docking. Does BOARD for a same-government ship with free bays mean; BOARD to refuel it, or BOARD to dock with it? For AI-controlled fighters we choose in the past to just don't allow boarding to resolve this ambiguity. Player controlled fighters now also follow the same behavior.

For the future I feel we should split the BOARD command in a DOCK and a BOARD command, but we are not there yet.

Comment (quyykk) [reply source/Engine.cpp:1635]: That sounds reasonable, although as it currently is written it will also disallow boarding different-government ships (like a player flying a fighter that wants to board a disabled pirate, that should never be interpreted as docking though right?).

Comment (petervdmeer) [reply source/Engine.cpp:1635]: Which is very relevant for https://github.com/endless-sky/endless-sky/pull/6323 where I probably copied this code from, but it seems less relevant for this specific PR.

I'll check if there still is a reason to have this code here as well. (And if not, then I'll remove it here.)

Comment (petervdmeer) [reply source/Engine.cpp:1635]: I checked `Ship::Board()`, and that function is also very explicit in that ships which have `CanBeCarried()` return true can only dock and not board.
My preference would be to not spend any additional effort here to create exceptions for fighter flagships, but to address fighter boarding more genericly in some follow-up PR.

Comment (quyykk) [reply source/Engine.cpp:1635]: Oh I see. The original behavior sets (for fighter flagships) CanBeCarried to false (through AllowCarried), which lets fighter flagships board (and not dock).  So that is no longer the case then. Follow PR is fine by me, but that would mean that people using fighter flagships aren't going to be able to board ships anymore until then.

Comment (petervdmeer) [reply source/Engine.cpp:1635]: Agreed, and I already created issue https://github.com/endless-sky/endless-sky/issues/5338 to request the feature to allow (some) fighters to dock. (And to disallow docking for some other ships.)

**Comment by quyykk** [comment]
The only thing missing for this PR is to restore the special exception that exists for flagship fighters. Since this bugfix is pretty good I think it's a good idea to just implement this exception (which would just be an additional parameter if I'm not wrong) and get this in 0.915.

We can always add separate docking and boarding commands later.

File: source/PlayerInfo.cpp [hunk 10]
```diff
@@ -49,6 +49,26 @@ PARTICULAR PURPOSE.  See the GNU General Public License for more details.
 
 using namespace std;
 
+namespace {
+	// Move the flagship to the start of your list of ships. It does not make sense
+	// that the flagship would change if you are reunited with a different ship that
+	// was higher up the list.
+	// This function should not be used while other functions have direct references
+	// to the shared_pointer elements in this list.
```
[the remark this answers is not in the document]
Comment (petervdmeer) [reply source/PlayerInfo.cpp:57]: Agreed, if some outside code just stores the shared-pointer (or the raw pointer) itself, then there is no issue.
The only issue which I was worried about is outside code storing a reference to the shared-pointer, but that would already be quite a bad practice so I guess I can just remove the comment.

File: source/PlayerInfo.cpp [hunk 11]
```diff
@@ -49,6 +49,26 @@ PARTICULAR PURPOSE.  See the GNU General Public License for more details.
...
+	void MoveFlagshipBegin(vector<shared_ptr<Ship>> &ships, const shared_ptr<Ship> &flagship)
+	{
+		if(!flagship)
+			return;
+
+		auto it = find(ships.begin(), ships.end(), flagship);
+		if(it != ships.begin() && it != ships.end())
+		{
+			move_backward(ships.begin(), it, next(it));
+			ships[0] = flagship;
```
[the remark this answers is not in the document]
Comment (petervdmeer) [reply source/PlayerInfo.cpp:67]: I also had to look up the arguments for `move_backward(...)` when quyykk suggested this, but I feel the usage of container code is actually quite elegant here. I'll add some comments to see if it improves the understandabilitly of the code.

File: source/PlayerInfo.cpp [hunk 12]
```diff
@@ -1266,30 +1330,10 @@ bool PlayerInfo::TakeOff(UI *ui)
...
-		shared_ptr<Ship> ship = it->lock();
-		if(!ship || ship == flagship)
-			it = selectedShips.erase(it);
-		else
-			++it;
-	}
+	// Move the flagship to the start of the list of ships and ensure that all
+	// escorts know which ship is acting as flagship.
+	MoveFlagshipBegin(ships, flagship);
+	SetFlagship(*flagship);
```
[the remark this answers is not in the document]
Comment (petervdmeer) [reply source/PlayerInfo.cpp:1338]: Maybe, but the re-ordering was originally only done at take-off and at landing, while SetFlagship can also be used during flight. I didn't check yet if doing the re-ordering during flight would be safe.

File: source/PlayerInfo.cpp [hunk 13]
```diff
@@ -795,6 +815,45 @@ const shared_ptr<Ship> &PlayerInfo::FlagshipPtr()
 
 
 
+// Set the flagship (on departure or during flight).
+void PlayerInfo::SetFlagship(Ship &other)
+{
+	// Remove active data in the old flagship.
```
[the remark this answers is not in the document]
Comment (petervdmeer) [reply source/PlayerInfo.cpp:823]: Making the method private seems to be sufficient for this PR.

File: source/PlayerInfo.cpp (hunk 12 above)
[the remark this answers is not in the document]
Comment (petervdmeer) [reply source/PlayerInfo.cpp:1338]: Which is not that relevant for this PR, because this PR doesn't deal with the in-flight part...

File: source/PlayerInfo.cpp [hunk 14]
```diff
@@ -1151,6 +1215,10 @@ void PlayerInfo::Land(UI *ui)
 			++it;
 	}
 
+	// If we switched flagships during flight, then we want to move the new
+	// flagship to the start of the list of ships.
```
[the remark this answers is not in the document]
Comment (petervdmeer) [reply source/PlayerInfo.cpp:1221]: Agreed, I will remove this code.

File: source/PlayerInfo.cpp (hunk 12 above)
[the remark this answers is not in the document]
Comment (petervdmeer) [reply source/PlayerInfo.cpp]: Moved the responsiblity for this to within SetFlagship.

File: source/PlayerInfo.cpp [hunk 15]
```diff
@@ -795,6 +815,45 @@ const shared_ptr<Ship> &PlayerInfo::FlagshipPtr()
...
+	for(const shared_ptr<Ship> &ship : ships)
+	{
+		bool shouldHaveParent = (flagship && ship != flagship && !ship->IsParked() && (!ship->CanBeCarried() || ship->JumpFuel()));
+		ship->SetParent(shouldHaveParent ? flagship : shared_ptr<Ship>());
+	}
+
+	// Make sure your flagship is not included in the escort selection.
+	// Or if we have no flagship, then make sure no escorts are selected.
+	if(!flagship)
+		selectedShips.clear();
```
[the remark this answers is not in the document]
Comment (petervdmeer) [reply source/PlayerInfo.cpp]: Agreed, Removed the if-part.

File: source/PlayerInfo.cpp [hunk 16]
```diff
@@ -822,6 +881,11 @@ map<const shared_ptr<Ship>, vector<string>> PlayerInfo::FlightCheck() const
 				flightChecks.emplace(ship, checks);
 
 			categoryCount[ship->Attributes().Category()].emplace_back(ship);
+			// Ensure bayCount has an entry for this category for the special case
+			// where we have no bays at all available for this type of ship.
+			if(ship->CanBeCarried())
+				bayCount[ship->Attributes().Category()];
```
[the remark this answers is not in the document]
Comment (petervdmeer) [reply source/PlayerInfo.cpp]: Switched to use emplace to make more clear that this is not about access, but about ensuring the element exists.

**Review by tehhowch — COMMENTED** [review]
A small nit, otherwise this looks fine

File: source/Engine.cpp (hunk 9 above)
Comment (petervdmeer) [reply source/Engine.cpp]: Just tested; the fighter does seem to start boarding, but doesn't seem to finish. There is probably still is a check somewhere that prevents boarding in this case.

**Comment by Hecter94** [comment]
@Amazinite This looks good to merge. Tehhowch has given it approval at this point since his one small nit was addressed exactly as he asked, and it's milestoned for 0.9.16, so it would be good to get it merged. 
I see no reason to wait any longer.

File: source/Engine.cpp (hunk 9 above)
Comment (petervdmeer) [reply source/Engine.cpp]: Those lines seem to prevent the boarding: https://github.com/endless-sky/endless-sky/blob/58b581f6f0ff199600e46994903cc8d812f87be3/source/Ship.cpp#L1996
